src/game-state.ts

```diff
@@ -0,0 +1,152 @@
+import { ChatCompletionMessageParam } from "openai/resources/index.mjs";
+import { Player, PrivateInfo, UserInteraction } from "./types";
+import crypto from 'crypto';
+import { ChatService, DefaultOpenAIClient } from "./openai";
+import { List, Map } from "immutable";
+
+/**
+ * Game state is immutable, using persistent data structures for efficient copying.
+ * Methods that update game state return a new instance of the game state. The caller is responsible for storing the new instance.
+ * See: https://github.com/immutable-js/immutable-js#the-case-for-immutability
+ * This creates a consistent interface between sync/async updates, avoiding mutation while awaiting a promise.
+ * 
+ * Where values are not expected to evolve, native (mutable) types may be used for convenience.
+ * If data is mutated inside of immutable structure, behavior is undefined.
+ */
+
+const openAIClient = new DefaultOpenAIClient(process.env.OPENAI_API_KEY || "");
+const openAIService = new ChatService(openAIClient);
+
+export class ScenarioState {
+  readonly canon: List<ChatCompletionMessageParam>;
+  readonly privateInfo: PrivateInfo;
+  readonly hash: string;
+  readonly parentHash: string | undefined;
+
+  protected constructor(canon: List<ChatCompletionMessageParam>, privateInfo: PrivateInfo, parentHash: string | undefined) {
+    this.canon = canon;
+    this.privateInfo = privateInfo;
+    this.parentHash = parentHash;
+    const stateString = JSON.stringify({
+      canon: this.canon.toArray(),
+      privateInfo: this.privateInfo,
+      parentHash: this.parentHash
+    });
+    this.hash = crypto.createHash('sha256').update(stateString).digest('hex').slice(0, 8);
```

Comment:
if we're going to go automated, let's actually remove the slicing and keep the whole hash